README.md
# numbersBR

[![Travis build status](https://travis-ci.org/wilsonfreitas/numbersBR.svg?branch=master)](https://travis-ci.org/wilsonfreitas/numbersBR)

The numbersBR package aims to help with brazilian numbers used to identify companies, individuals and cars (CNPJ, CPF and RENAVAN, for example).

The package provides functions to validate, format and execute comparisons of these numbers when they come in different formats.
For example, compare a numeric CPF against a character CPF.
The functions are implemented in a vectorized way in order to speed up validations and comparison in large datasets.

## Install

The package can be installed directly from github using devtools.

```{r}
devtools::install_github('wilsonfreitas/numbersBR')
```

## Using

The numbers can be created with numeric or character vectors:

```{r}
> x <- CNPJ(13515463000138)
> x
13.515.463/0001-38 
> is.valid(x)
[1] TRUE
> x == 13515463000138
[1] TRUE
> x <- CPF(c("681.943.594-06", "012.391.576-73", "520.082.755-82"))
> x
681.943.594-06 012.391.576-73 520.082.755-82 
> is.valid(x)
[1] TRUE TRUE TRUE
> x == 68194359406
[1]  TRUE FALSE FALSE
> x <- RENAVAN(c(75320797785, 91671999129))
> is.valid(x)
[1] TRUE TRUE
```

## Contribute

For now we have CPF, CNPJ and RENAVAN implemented, so if you need another number or, even better, if you have it implemented, get in touch so we can add new numbers to the package.


NEWS.md

Changes in Version 0.0.2

- First stable version

- Implements CPF, CNPJ, RENAVAN



tests/testthat.R
library(testthat)
library(rBR)

test_check("rBR")

tests/testthat/test-CPF.R
context('CPF')

test_that('it should instanciate a CPF', {
	x <- CPF("68194359406")
	expect_is(x, 'CPF')
	x <- CPF("681.943.594-06")
	expect_is(x, 'CPF')
	x <- CPF(68194359406)
	expect_is(x, 'CPF')
	x <- CPF(c(68194359406, 1239157673, 52008275582))
	expect_is(x, 'CPF')
})

test_that('it should not instanciate a CPF', {
	expect_error(CPF(NULL), 'Unsupported type')
})

test_that('it should check how CPF is stored internally', {
	x <- CPF("68194359406")
	expect_equal(unclass(x), "68194359406")
	x <- CPF(1239157673)
	expect_equal(unclass(x), "01239157673")
	x <- CPF(c("681.943.594-06", "012.391.576-73", "520.082.755-82"))
	expect_equal(unclass(x), c("68194359406", "01239157673", "52008275582"))
})

test_that('it should format CPF', {
	x <- CPF("681.943.594-06")
	expect_equal(format(x), "681.943.594-06")
	x <- CPF(1239157673)
	expect_equal(format(x, "strict"), "012.391.576-73")
	expect_equal(format(x, "stripped"), "01239157673")
	x <- CPF(c(68194359406, 1239157673, 52008275582))
	expect_equal(format(x), c("681.943.594-06", "012.391.576-73", "520.082.755-82"))
})

test_that('it should convert CPF into numeric', {
	x <- CPF("681.943.594-06")
	expect_equal(as.numeric(x), 68194359406)
	x <- CPF("01239157673")
	expect_equal(as.numeric(x), 1239157673)
	x <- CPF(c("681.943.594-06", "012.391.576-73", "520.082.755-82"))
	expect_equal(as.numeric(x), c(68194359406, 1239157673, 52008275582))
})

test_that('it should convert CPF into character', {
	x <- CPF(68194359406)
	expect_equal(as.character(x), "681.943.594-06")
	x <- CPF(1239157673)
	expect_equal(as.character(x), "012.391.576-73")
	x <- CPF(c(68194359406, 1239157673, 52008275582))
	expect_equal(as.character(x), c("681.943.594-06", "012.391.576-73", "520.082.755-82"))
})

test_that('it should CPF`s equality', {
	x <- CPF(68194359406)
	expect_true(x == 68194359406)
	expect_true(x == "681.943.594-06")
	expect_false(x != "681.943.594-06")
	expect_false(x == "012.391.576-73")
	expect_true(x != 1239157673)
	x <- CPF(c(68194359406, 1239157673, 52008275582))
	expect_true(all( x == c("681.943.594-06", "012.391.576-73", "520.082.755-82") ))
})

test_that('it should validate CPF', {
	x <- CPF(68194359406)
	expect_true(is.valid(x))
	x <- CPF(1239157673)
	expect_true(is.valid(x))
	x <- CPF(42752486198)
	expect_false(is.valid(x))
	x <- CPF(c(68194359406, 1239157673))
	expect_true(all(is.valid(x)))
	x <- CPF(c(1239157673, 42752486198))
	expect_equal(is.valid(x), c(T, F))
})



tests/testthat/test-CNPJ.R
context('CNPJ')

test_that('it should instanciate a CNPJ', {
	x <- CNPJ("13.515.463/0001-38")
	expect_is(x, 'CNPJ')
	x <- CNPJ("13515463000138")
	expect_is(x, 'CNPJ')
	x <- CNPJ(13515463000138)
	expect_is(x, 'CNPJ')
	x <- CNPJ(c("13.515.463/0001-38", "03.737.211/0001-08",
		"00.360.305/0001-04", "00.066.670/0001-00"))
	expect_is(x, 'CNPJ')
})

test_that('it should not instanciate a CNPJ', {
	expect_error(CNPJ(NULL), 'Unsupported type')
})

test_that('it should check how CNPJ is stored internally', {
	x <- CNPJ("13.515.463/0001-38")
	expect_equal(unclass(x), "13515463000138")
	x <- CNPJ(13515463000138)
	expect_equal(unclass(x), "13515463000138")
	x <- CNPJ(66670000100)
	expect_equal(unclass(x), "00066670000100")
	x <- CNPJ(c("13.515.463/0001-38", "03.737.211/0001-08",
		"00.360.305/0001-04", "00.066.670/0001-00"))
	expect_equal(unclass(x), c("13515463000138", "03737211000108",
		"00360305000104", "00066670000100"))
})

test_that('it should format CNPJ', {
	x <- CNPJ("13.515.463/0001-38")
	expect_equal(format(x), "13.515.463/0001-38")
	x <- CNPJ(66670000100)
	expect_equal(format(x), "00.066.670/0001-00")
	x <- CNPJ(c(13515463000138, 3737211000108, 360305000104, 66670000100))
	expect_equal(format(x), c("13.515.463/0001-38", "03.737.211/0001-08",
		"00.360.305/0001-04", "00.066.670/0001-00"))
})

test_that('it should convert CNPJ into numeric', {
	x <- CNPJ("13.515.463/0001-38")
	expect_equal(as.numeric(x), 13515463000138)
	x <- CNPJ("00.066.670/0001-00")
	expect_equal(as.numeric(x), 66670000100)
	x <- CNPJ(c("13.515.463/0001-38", "03.737.211/0001-08",
		"00.360.305/0001-04", "00.066.670/0001-00"))
	expect_equal(as.numeric(x), c(13515463000138, 3737211000108, 360305000104, 66670000100))
})

test_that('it should convert CNPJ into character', {
	x <- CNPJ(13515463000138)
	expect_equal(as.character(x), "13.515.463/0001-38")
	x <- CNPJ(66670000100)
	expect_equal(as.character(x), "00.066.670/0001-00")
	x <- CNPJ(c(13515463000138, 3737211000108, 360305000104, 66670000100))
	expect_equal(as.character(x), c("13.515.463/0001-38", "03.737.211/0001-08",
		"00.360.305/0001-04", "00.066.670/0001-00"))
})

test_that('it should validate CNPJ', {
	x <- CNPJ(13515463000138)
	expect_true(is.valid(x))
	x <- CNPJ(66670000100)
	expect_true(is.valid(x))
	x <- CNPJ(66670000101)
	expect_false(is.valid(x))
	x <- CNPJ(c(13515463000138, 3737211000108, 360305000104, 66670000100))
	expect_true(all(is.valid(x)))
	x <- CNPJ(c(13515463000138, 66670000101))
	expect_equal(is.valid(x), c(T, F))
})

test_that('it should CNPJ`s equality', {
	x <- CNPJ(13515463000138)
	expect_true(x == 13515463000138)
	expect_true(x == "13.515.463/0001-38")
	expect_false(x != "13.515.463/0001-38")
	expect_false(x == 66670000100)
	expect_true(x != 66670000100)
	x <- CNPJ(c(13515463000138, 3737211000108, 360305000104, 66670000100))
	expect_true(all( x == c("13.515.463/0001-38", "03.737.211/0001-08",
		"00.360.305/0001-04", "00.066.670/0001-00") ))
})



R/CPF.R
#' @export
#' @rdname numbers
#' @examples
#' CPF(c(68194359406, 1239157673, 52008275582))
#' CPF(c("681.943.594-06", "012.391.576-73", "520.082.755-82"))
CPF <- function(x) {
  if ( is.character(x) ) {
    x <- stringr::str_replace_all(x, '-', '')
    x <- stringr::str_replace_all(x, '\\.', '')
    x <- stringr::str_pad(x, 11, pad="0")
  } else if ( is.numeric(x) ) {
    x <- as.character(x)
    x <- stringr::str_pad(x, 11, pad="0")
  } else {
    stop('Unsupported type ', class(x))
  }

  structure(x, class='CPF')
}

#' @export
#' @rdname format
#' @examples
#' x <- CPF(1239157673)
#' format(x)
#' format(x, "stripped")
format.CPF <- function(x, format = c("strict", "stripped"), ...) {
  format = match.arg(format)
  x <- unclass(x)
  if (format == "strict")
    sprintf("%s.%s.%s-%s",
            stringr::str_sub(x, 1, 3),
            stringr::str_sub(x, 4, 6),
            stringr::str_sub(x, 7, 9),
            stringr::str_sub(x, 10, 11))
  else
    as.character(x)
}

#' @export
as.character.CPF <- function(x, ...) {
  format(x)
}

#' @export
as.numeric.CPF <- function(x, ...) {
  x <- unclass(x)
  as.numeric(x)
}

#' @export
#' @rdname is.valid
#' @examples
#' is.valid(CPF(c(1239157673, 42752486198))) # TRUE, FALSE
is.valid.CPF <- function(x) {
  x1 <- as.numeric(x)

  a <-     1*extract_digit(x1, 11)
  a <- a + 2*extract_digit(x1, 10)
  a <- a + 3*extract_digit(x1, 9)
  a <- a + 4*extract_digit(x1, 8)
  a <- a + 5*extract_digit(x1, 7)
  a <- a + 6*extract_digit(x1, 6)
  a <- a + 7*extract_digit(x1, 5)
  a <- a + 8*extract_digit(x1, 4)
  a <- a + 9*extract_digit(x1, 3)
  a <- (a %% 11) %% 10

  b <-     1*extract_digit(x1, 10)
  b <- b + 2*extract_digit(x1, 9)
  b <- b + 3*extract_digit(x1, 8)
  b <- b + 4*extract_digit(x1, 7)
  b <- b + 5*extract_digit(x1, 6)
  b <- b + 6*extract_digit(x1, 5)
  b <- b + 7*extract_digit(x1, 4)
  b <- b + 8*extract_digit(x1, 3)
  b <- b + 9*extract_digit(x1, 2)
  b <- (b %% 11) %% 10

  a == extract_digit(x1, 2) & b == extract_digit(x1, 1)
}

#' @export
#' @rdname comparison-methods
#' @examples
#' x <- CPF(68194359406)
#' x != "681.943.594-06"
`==.CPF` <- function(e1, e2) {
  unclass(e1) == unclass(CPF(e2))
}

#' @export
#' @rdname comparison-methods
`!=.CPF` <- function(e1, e2) {
  !(e1 == e2)
}

#' @export
print.CPF <- function(x, ...) {
  x <- format(x)
  cat(x, '\n')
  invisible(x)
}


R/RENAVAN.R
#' @export
#' @rdname numbers
#' @examples
#' RENAVAN(c(75320797785, 91671999129, 60885589380))
RENAVAN <- function(x) {
  if ( is.character(x) ) {
    x <- stringr::str_pad(x, 11, pad="0")
  } else if ( is.numeric(x) ) {
    x <- as.character(x)
    x <- stringr::str_pad(x, 11, pad="0")
  } else {
    stop('Unsupported type ', class(x))
  }

  structure(x, class='RENAVAN')
}

#' @export
#' @rdname format
#' @examples
#' x <- RENAVAN("68194359406")
#' format(x)
format.RENAVAN <- function(x, ...) {
  x <- unclass(x)
  as.character(x)
}

#' @export
as.character.RENAVAN <- function(x, ...) {
  format(x)
}

#' @export
as.numeric.RENAVAN <- function(x, ...) {
  x <- unclass(x)
  as.numeric(x)
}

#' @export
#' @rdname is.valid
#' @examples
#' is.valid(RENAVAN(c(75320797785, 42752486198))) # TRUE, FALSE
is.valid.RENAVAN <- function(x) {
  check_renavan <- function(renavan) {
    renavan <- strip_numeric(renavan)
    control <- strip_numeric(3298765432)

    sum_ <- Reduce(sum, Map(function(x, y) x * y, renavan[-length(renavan)], control))

    digit_ <- 11L - sum_ %% 11L
    digit_ <- if (digit_ < 10L) digit_ else 0L

    renavan[length(renavan)] == digit_
  }
  renavan <- as.numeric(x)
  unlist( Map(check_renavan, renavan) )
}

#' @export
#' @rdname comparison-methods
#' @examples
#' x <- RENAVAN(68194359406)
#' x == "68194359406"
`==.RENAVAN` <- function(e1, e2) {
  unclass(e1) == unclass(RENAVAN(e2))
}

#' @export
#' @rdname comparison-methods
`!=.RENAVAN` <- function(e1, e2) {
  !(e1 == e2)
}

#' @export
print.RENAVAN <- function(x, ...) {
  x <- format(x)
  cat(x, '\n')
  invisible(x)
}


R/numbersBR.R
#' Validate, Compare and Format Identification Numbers From Brazil.
#'
#' This package helps handling identification numbers used in Brazil.
#' These numbers are used to identify individuals (CPF), vehicles (RENAVAN),
#' companies (CNPJ) and etc.
#' Functions to format, validate and compare these numbers have been
#' implemented in a vectorized way in order to speed up validations and
#' comparisons in big datasets.
#'
#' @author Wilson Freitas
#' @name numbersBR
#' @docType package
#'
#' @import stringr
NULL

#' Functions to create brazilian numbers objects
#'
#' The brazilian numbers used to identify individuals, vehicles, companies and
#' any other things have their own S3 classes.
#'
#' @param x the identification number as numeric or character
#'
#' @return A S3 class representing the brazilian number: CNPJ, CPF, RENAVAN
#'
#' @name numbers
NULL

#' Format numbers
#'
#' Formats the brazilian numbers according to their specific format.
#'
#' @param x the identification number class.
#' @param format can be strict which is the default format with all dots, dashes and hyphens or
#' stripped which returns only the numbers.
#' @param ... arguments to be passed to or from other methods.
#'
#' @return A character vector with the formatted number.
#'
#' @name format
NULL

#' Validate numbers
#'
#' Validates the brazilian numbers when it is necessary.
#' Not all identification numbers don't have a validation algorithm.
#'
#' @param x the identification number class.
#'
#' @return A logical vector indicating whether the number is valid or not.
#'
#' @name is.valid
NULL

#' Compare numbers among themselves, with numeric and character objects.
#'
#' Compares the brazilian numbers among themselves, with numeric and character objects.
#'
#' @param e1 the identification number class.
#' @param e2 the object to be compared with, can be of the same type of \code{e1}, a numeric or character vector
#'
#' @return A logical vector indicating that the given objects represent the same number.
#'
#' @name comparison-methods
NULL



R/CNPJ.R
#' @export
#' @rdname numbers
#' @examples
#' CNPJ(c(13515463000138, 3737211000108, 360305000104, 66670000100))
#' CNPJ(c("13.515.463/0001-38", "03.737.211/0001-08", "00.360.305/0001-04", "00.066.670/0001-00"))

CNPJ <- function(x) {
  if ( is.character(x) ) {
    x <- stringr::str_replace_all(x, '-', '')
    x <- stringr::str_replace_all(x, '\\.', '')
    x <- stringr::str_replace_all(x, '/', '')
    x <- stringr::str_pad(x, 14, pad="0")
  } else if ( is.numeric(x) ) {
    x <- as.character(x)
    x <- stringr::str_pad(x, 14, pad="0")
  } else {
    stop('Unsupported type ', class(x))
  }

  structure(x, class='CNPJ')
}

#' @export
#' @rdname format
#' @examples
#' x <- CNPJ(66670000100)
#' format(x)
#' format(x, "stripped")
format.CNPJ <- function(x, format = c("strict", "stripped"), ...) {
  format = match.arg(format)
  x = unclass(x)
  if (format == "strict")
    sprintf("%s.%s.%s/%s-%s",
            stringr::str_sub(x, 1, 2),
            stringr::str_sub(x, 3, 5),
            stringr::str_sub(x, 6, 8),
            stringr::str_sub(x, 9, 12),
            stringr::str_sub(x, 13, 14))
  else
    as.character(x)
}

#' @export
as.character.CNPJ <- function(x, ...) {
  format(x)
}

#' @export
as.numeric.CNPJ <- function(x, ...) {
  x <- unclass(x)
  as.numeric(x)
}

#' @export
#' @rdname is.valid
#' @examples
#' is.valid(CNPJ(c(13515463000138, 66670000101))) # TRUE, FALSE
is.valid.CNPJ <- function(x) {
  x <- unclass(x)
  x1 <- as.numeric(stringr::str_sub(x, 1, 12))
  v1 <- as.integer(stringr::str_sub(x, 13, 13))
  v2 <- as.integer(stringr::str_sub(x, 14, 14))

  a <- 5*extract_digit(x1, 12)
  a <- a + 4*extract_digit(x1, 11)
  a <- a + 3*extract_digit(x1, 10)
  a <- a + 2*extract_digit(x1, 9)
  a <- a + 9*extract_digit(x1, 8)
  a <- a + 8*extract_digit(x1, 7)
  a <- a + 7*extract_digit(x1, 6)
  a <- a + 6*extract_digit(x1, 5)
  a <- a + 5*extract_digit(x1, 4)
  a <- a + 4*extract_digit(x1, 3)
  a <- a + 3*extract_digit(x1, 2)
  a <- a + 2*extract_digit(x1, 1)
  a <- 11 - a %% 11
  a[a >= 10] <- 0

  b <- 6*extract_digit(x1, 12)
  b <- b + 5*extract_digit(x1, 11)
  b <- b + 4*extract_digit(x1, 10)
  b <- b + 3*extract_digit(x1, 9)
  b <- b + 2*extract_digit(x1, 8)
  b <- b + 9*extract_digit(x1, 7)
  b <- b + 8*extract_digit(x1, 6)
  b <- b + 7*extract_digit(x1, 5)
  b <- b + 6*extract_digit(x1, 4)
  b <- b + 5*extract_digit(x1, 3)
  b <- b + 4*extract_digit(x1, 2)
  b <- b + 3*extract_digit(x1, 1)
  b <- b + 2*v1
  b <- 11 - b %% 11
  b[b >= 10] <- 0

  a == v1 & b == v2
}

#' @export
#' @rdname comparison-methods
#' @examples
#' x <- CNPJ(13515463000138)
#' x == '13515463000138'
`==.CNPJ` <- function(e1, e2) {
  unclass(e1) == unclass(CNPJ(e2))
}

#' @export
#' @rdname comparison-methods
`!=.CNPJ` <- function(e1, e2) {
  !(e1 == e2)
}

#' @export
print.CNPJ <- function(x, ...) {
  x <- format(x)
  cat(x, '\n')
  invisible(x)
}
